logic/path/src/selectors/legacy.rs

```diff
@@ -75,73 +107,134 @@ impl<CW> DfsPathSelector<CW>
 where
     CW: EdgeWeighting<U256>,
 {
+    /// Determines whether a node is considered "interesting".
+    ///
+    /// To achieve privacy, we need at least one honest node along
+    /// the path. Each additional node decreases the probability of
+    /// having only malicious nodes, so we can sort out many nodes.
+    /// Nodes that have shown to be unreliable are of no use, so
+    /// drop them.
     fn filter_channel(
         &self,
         channel: &ChannelEdge,
+        source: &Address,
         destination: &Address,
-        current: &[Address],
-        dead_ends: &HashSet<Address>,
+        current_path: &[Address],
     ) -> bool {
-        !destination.eq(&channel.channel.destination) &&                    // last hop does not need a channel
-            channel.quality.unwrap_or(1_f64) > self.quality_threshold &&  // quality threshold
-            !current.contains(&channel.channel.destination) &&     // must not be in the path already (no loops allowed)
-            !dead_ends.contains(&channel.channel.destination) // must not be in the dead end list
+        if source.eq(&channel.channel.destination) {
+            // looping back to self does not give any privacy
+            return false;
+        }
+
+        if destination.eq(&channel.channel.destination) {
+            // We cannot use destination as last intermediate hop as
+            // this would be a loopback which does not give any privacy
+            return false;
+        }
+
+        // Check node reliability, new nodes are considered reliable
+        // unless the opposite has been observed
+        if channel.quality.unwrap_or(1.0f64) < self.quality_threshold {
+            // Only use nodes that have shown to be somewhat reliable
+            return false;
+        }
+
+        if current_path.contains(&channel.channel.destination) {
+            // At the moment, we do not allow circles because they
+            // do not give additional privacy
+            return false;
+        }
+
+        if U256::zero().eq(&channel.channel.balance.amount()) {
+            // We cannot use channels with zero stake
+            return false;
+        }
```

Comment:
This should be configurable. The risk of a fractured network in this case means there might be lower privacy due to the inability of passing through the existing transport links without any balance.

Reply:
Privacy first or earning first?

Reply:
made it configurable